### PR TITLE
Update to use new data structure

### PR DESCRIPTION
**Pull request by semanticart**
There's some open TODOs to resolve here. I'll add some notes.

### CODE REVIEW

File: lib/prefab/config_value_unwrapper.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def self.unwrap(config_value, config_key, properties)
+      return nil unless config_value
+
+      case config_value.type
+      when :int, :string, :double, :bool, :log_level
+        config_value.public_send(config_value.type)
+      when :string_list
+        config_value.string_list.values
+      when :bytes
+        raise 'TODO: bytes'
```
Comment (semanticart) [inline lib/prefab/config_value_unwrapper.rb:14]: I'm not sure we need to handle this or `limit_definition` yet. If we don't, I'm happy to delete them for now.

File: lib/prefab/criteria_evaluator.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    def initialize(config, project_env_id:, resolver:, base_client:)
+      @config = config
+      @project_env_id = project_env_id
+      @resolver = resolver
+      @base_client = base_client
+    end
+
+    def evaluate(properties)
+      # TODO: optimize this and perhaps do it elsewhere
+      props = properties.transform_keys(&:to_s)
```
Comment (semanticart) [inline lib/prefab/criteria_evaluator.rb:18]: We can punt this to a future PR

File: lib/prefab/criteria_evaluator.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+      matching_environment_row_values.each do |conditional_value|
+        # TODO: merge in properties from the environment_row?
```
Comment (semanticart) [inline lib/prefab/criteria_evaluator.rb:21]: @jdwyah should we handle this now or later?

File: lib/prefab/config_value_unwrapper.rb (hunk 1 above)
Comment (jdwyah) [reply lib/prefab/config_value_unwrapper.rb:14]: Yeah, I don't think either of them is near term roadmap.

File: lib/prefab/criteria_evaluator.rb (hunk 3 above)
Comment (jdwyah) [reply lib/prefab/criteria_evaluator.rb:21]: Later ok imo.

File: lib/prefab/config_value_unwrapper.rb (hunk 1 above)
Comment (semanticart) [reply lib/prefab/config_value_unwrapper.rb:14]: Perfect. Axing for now.

File: prefab-cloud-ruby.gemspec [hunk 4]
```diff
@@ -38,31 +40,43 @@ Gem::Specification.new do |s|
     "lib/prefab/config_helper.rb",
```
Comment (mikowitz) [inline prefab-cloud-ruby.gemspec:40]: We can remove `config_helper.rb`, I think.

File: lib/prefab/criteria_evaluator.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        value_from_properites.start_with?(criterion.value_to_match.string)
+      else
+        @base_client.log.info("Unknown Operator: #{criteria.operator}")
+        false
+      end
+    end
+
+    private
+
+    def matches?(criterion, value_from_properites, properties)
```
Comment (mikowitz) [inline lib/prefab/criteria_evaluator.rb:73]: code organization 🎱 : Maybe move this down to be next to `in_segment?` so the evaluation helpers are together?

File: lib/prefab/feature_flag_client.rb [hunk 6]
```diff
@@ -2,53 +2,32 @@
...
 
-    def upsert(feature_name, feature_obj)
-      @base_client.config_client.upsert(feature_name, Prefab::ConfigValue.new(feature_flag: feature_obj))
-    end
-
     def feature_is_on?(feature_name)
       feature_is_on_for?(feature_name, nil)
     end
 
     def feature_is_on_for?(feature_name, lookup_key, attributes: {})
```
Comment (mikowitz) [inline lib/prefab/feature_flag_client.rb:13]: 🎱 should we rename them `properties` here too, like in the config client, while we're here?

**Review by mikowitz — COMMENTED** [review]
Overall this looks really good! A couple questions, mostly of the 🎱 variety, both nothing major.

Are we saving provenance for a future PR, or do we not need that explicit of a concept in the client?

File: lib/prefab/client.rb [hunk 7]
```diff
@@ -97,18 +97,22 @@ def enabled?(feature_name, lookup_key = nil, attributes = {})
       feature_flag_client.feature_is_on_for?(feature_name, lookup_key, attributes: attributes)
```
Comment (mikowitz) [inline lib/prefab/client.rb:97]: 🎱  Should we replace `attributes` with `properties` here too, while we're here?

**Comment by semanticart** [comment]
> Are we saving provenance for a future PR, or do we not need that explicit of a concept in the client?

Saving this one for a future PR.

File: lib/prefab/config_client.rb [hunk 8]
```diff
@@ -74,18 +72,18 @@ def self.value_to_delta(key, config_value, namespace = nil)
...
 
-    def get(key, default = Prefab::Client::NO_DEFAULT_PROVIDED)
-      config = _get(key)
-      config ? value_of(config[:value]) : handle_default(key, default)
-    end
-
-    def get_config_obj(key)
-      config = _get(key)
-      config ? config[:config] : nil
+    # TODO: this call signature has changed, update docs
```
[the remark this answers is not in the document]
Comment (semanticart) [reply lib/prefab/config_client.rb:75]: Oooh, that would be great, actually. Ping me on the PR, please.

File: test/test_integration.rb [hunk 9]
```diff
@@ -11,6 +11,11 @@ class TestIntegration < Minitest::Test
 
     tests.each do |test|
       define_method(:"test_#{test['name']}") do
+        puts "TODO: #{test}"
+
+        # TODO: handle this
```
[the remark this answers is not in the document]
Comment (semanticart) [reply test/test_integration.rb:16]: Correct!

File: prefab-cloud-ruby.gemspec (hunk 4 above)
Comment (semanticart) [reply prefab-cloud-ruby.gemspec]: Yes! f979b50

File: lib/prefab/feature_flag_client.rb (hunk 6 above)
Comment (semanticart) [reply lib/prefab/feature_flag_client.rb:13]: Good question. I have an open question for @jdwyah about standardizing this.

File: lib/prefab/criteria_evaluator.rb (hunk 5 above)
Comment (semanticart) [reply lib/prefab/criteria_evaluator.rb]: No strong feelings here. Tie goes to the runner 5654172

**Comment by mikowitz** [comment]
This all looks good to me, as long as the follow-up PR work is enumerated somewhere so it doesn't get lost, and staging is in a state this can actually talk to, I say :shipit:

File: lib/prefab/config_client.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (mikowitz) [reply lib/prefab/config_client.rb:75]: Oh, actually, I'm not sure the docs need this, at least not in https://github.com/prefab-cloud/prefab-cloud-docs-web/ Looks like the docs there only talk about `Client#get`, whose signature hasn't changed.

[the remark this answers is not in the document]
Comment (mikowitz) [reply lib/prefab/config_client.rb:75]: There are some code samples in `bullet-web` that make use of `config_client.get` explicitly, so those will need to be updated, but then the question is should we update them to still use the config client, or have them match what's in `docs-web`?

[the remark this answers is not in the document]
Comment (semanticart) [reply lib/prefab/config_client.rb:75]: Ooh, good call! Let's update any code samples that reference internal clients (config_client or feature_flag_client).

**Comment by mikowitz** [comment]
That was just me rebasing after merging #70

**Pull request merged**